File: Frontend/tehtavaappfrontend/src/interfaces/models/Course.ts
import { SchoolGroup } from './SchoolGroup';
import { Block } from '../../types/blocks';

/**
 * Enhanced Course interface with all required properties
 * This contains a superset of all course properties used in the application
 */
export interface Course {
  id: string;
  name: string;
  description: string;
  code?: string;
  teacherId: string;
  teacherName?: string;
  createdById?: string;
  createdByName?: string;
  createdAt: string | Date;
  updatedAt?: string | Date;
  contentBlocks?: Block[];
  studentCount?: number;
  materialCount?: number;
  assignmentCount?: number;
  groups?: SchoolGroup[];
  startDate?: string | Date;
  endDate?: string | Date;
  isActive?: boolean;
}

/**
 * Interface for course enrollment
 */
export interface CourseEnrollment {
  id: string;
  courseId: string;
  userId: string;
  enrollmentDate: string | Date;
  status: string;
  completionDate?: string | Date;
  progress?: number;
}

/**
 * Interface for course teacher
 */
export interface CourseTeacher {
  id: string;
  courseId: string;
  teacherId: string;
  teacherName?: string;
  email?: string;
  isMainTeacher?: boolean;
  assignedDate?: string | Date;
} 

File: Backend/TehtavaApp.API/Models/Course.cs
namespace TehtavaApp.API.Models
{
    public class Course
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string Code { get; set; }
        public string TeacherId { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public bool IsActive { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        public string ContentBlocksJson { get; set; }

        // Navigation properties
        public ApplicationUser Teacher { get; set; }
        public ICollection<Assignment> Assignments { get; set; }
        public ICollection<Material> Materials { get; set; }
        public ICollection<SchoolGroup> Groups { get; set; }
        public ICollection<CourseTeacher> CourseTeachers { get; set; }

        public Course()
        {
            Assignments = new HashSet<Assignment>();
            Materials = new HashSet<Material>();
            Groups = new HashSet<SchoolGroup>();
            CourseTeachers = new HashSet<CourseTeacher>();
            CreatedAt = DateTime.UtcNow;
            UpdatedAt = DateTime.UtcNow;
            IsActive = true;
        }
    }
}
